lib/Dialect/SV/Transforms/GeneratorCallout.cpp

```diff
@@ -0,0 +1,159 @@
+//===- RTLLegalizeNames.cpp - RTL Name Legalization Pass ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Call arbitrary programs and pass them the attributes attached to external
+// modules.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVAnalyses.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Support/ImplicitLocOpBuilder.h"
+#include "circt/Translation/ExportVerilog.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Process.h"
+
+using namespace circt;
+using namespace sv;
+using namespace rtl;
+
+//===----------------------------------------------------------------------===//
+// GeneratorCalloutPass
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct RTLGeneratorCalloutPass
+    : public sv::RTLGeneratorCalloutPassBase<RTLGeneratorCalloutPass> {
+  void runOnOperation() override;
+
+private:
+  DenseMap<StringRef, RTLGeneratorSchemaOp> generatorSchema;
+};
+} // end anonymous namespace
+
+// Parse the \p genExec string, to extract the program executable, the path to
+// the executable and all the required arguments from \p genExecArgs.
+void static parseProgramArgs(const std::string genExec,
+                             const std::string genExecArgs,
+                             std::string &execName,
+                             std::string &genExecutablePath,
+                             std::vector<std::string> &genOptions) {
+  std::string temp = genExecArgs;
+  size_t pos;
+  while ((pos = temp.find_last_of(";")) != std::string::npos) {
+    genOptions.push_back(temp.substr(pos + 1));
+    temp = temp.substr(0, pos);
+  }
+  genOptions.push_back(temp);
+  temp = genExec;
+
+  auto pathLoc = temp.find_last_of("/\\");
+  if (pathLoc != std::string::npos)
+    genExecutablePath = temp.substr(0, pathLoc + 1);
+  else
+    genExecutablePath = "";
+  execName = temp.substr(pathLoc + 1);
+  return;
+}
+
+void RTLGeneratorCalloutPass::runOnOperation() {
+  ModuleOp root = getOperation();
+  std::string genExecutableName, genExecutablePath;
+  std::vector<std::string> genOptions;
+  parseProgramArgs(genExecutable, genExecArgs, genExecutableName,
+                   genExecutablePath, genOptions);
+
+  OpBuilder builder(root->getContext());
+  auto genMemExe =
```

Comment:
Deal with the executable in the executable finding function you split off from the parse function.